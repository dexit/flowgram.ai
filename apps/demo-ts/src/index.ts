export * from './types'
export * from './runtime'
export * from './translate'
export * from './validate'
export * from './util'
