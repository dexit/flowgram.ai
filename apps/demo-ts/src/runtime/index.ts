export * from './run'
export * from './RuntimeError'
