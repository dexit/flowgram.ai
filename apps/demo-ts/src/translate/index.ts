export * from './nodes'
export * from './translate'
