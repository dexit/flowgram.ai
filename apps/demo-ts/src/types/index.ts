export * from '@flow-ide-editor/schema'
export * from './nodes'

export class InvalidSchemaError extends Error {}
export class InvalidNodeTypeError extends Error {}
export class InvalidTypeError extends Error {}
export class NotImplementedError extends Error {}

export type FlowArgs = Record<string, unknown>
