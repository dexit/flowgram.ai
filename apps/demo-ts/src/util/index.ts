export * from './log'
export * from './JSBuffer'
