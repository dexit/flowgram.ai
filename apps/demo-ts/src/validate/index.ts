import { FlowNodeSchema } from '@flow-ide-editor/schema'
import { InvalidSchemaError } from '../types'

// TODO use json-schema validation
export function validateOrThrow(node: FlowNodeSchema): void {
  if (!node.definition) throw new InvalidSchemaError()
  if (!node.definition.info) throw new InvalidSchemaError()
  if (!node.definition.info.title) throw new InvalidSchemaError()
}
