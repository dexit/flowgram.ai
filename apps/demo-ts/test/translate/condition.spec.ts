// import _ from 'lodash'
import _ from 'lodash'
import { FlowConditionNodeSchema, FlowNodeSchema, translateNode } from '../../src'

const variable1: FlowNodeSchema = {
  id: 'variable_e0c4615094f',
  type: 'variable',
  definition: {
    info: { title: 'variable_1' },
    outputs: { properties: {}, title: 'Output', type: 'object' },
    script: { language: 'python', content: '' },
  },
  data: {
    setVariables: {
      '0': {
        left: { type: 'ref', content: '$.variable.outputs.a' },
        right: { type: 'expression', content: "'hello world'" },
        operator: 'assign',
      },
    },
  },
}

const variable2: FlowNodeSchema = {
  id: 'variable_dd605e4e286',
  type: 'variable',
  definition: {
    info: { title: 'variable_2', description: '', icon: '' },
    outputs: {
      type: 'object',
      properties: {
        a: { type: 'string' },
      },
    },
    script: { language: 'python', content: '' },
  },
  data: {
    setVariables: {
      '0': {
        left: { type: 'ref', content: '@.a' },
        right: { type: 'expression', content: 'start.outputs.query' },
        operator: 'assign',
      },
    },
  },
}

const conditionMin = {
  id: 'condition_4e0c4615094',
  type: 'condition',
  definition: {
    info: { title: 'condition' },
    script: { language: 'python', content: '' },
  },
  blocks: [
    {
      id: 'branch_f7fdf2437b3',
      type: 'case',
      definition: {
        info: { title: 'case_hello' },
      },
      data: {
        condition: { type: 'expression', content: "variable.outputs.a == 'hello'" },
      },
      blocks: [],
    },
    {
      id: 'branch_f7fdf2437b4',
      type: 'case',
      definition: {
        info: { title: 'case_hello1' },
      },
      data: {
        condition: { type: 'expression', content: "variable.outputs.a == 'hello1'" },
      },
      blocks: [],
    },
    {
      id: 'default_64e0c461509',
      type: 'default',
      definition: {
        info: {
          title: 'case_default',
          description: 'Executed when all conditions are not met',
          icon: '',
        },
      },
      blocks: [],
    },
  ],
}

describe('translate', () => {
  it('Condition - min', () => {
    const node = _.cloneDeep(conditionMin) as FlowConditionNodeSchema

    expect(translateNode(node)).toMatchSnapshot()
  })

  it('Condition - case/case/default', () => {
    const node: FlowConditionNodeSchema = {
      id: 'condition_4e0c4615094',
      type: 'condition',
      definition: {
        info: { title: 'condition' },
        script: { language: 'python', content: '' },
      },
      blocks: [
        {
          id: 'branch_f7fdf2437b3',
          type: 'case',
          definition: {
            info: { title: 'case_hello' },
          },
          data: {
            condition: { type: 'expression', content: "variable.outputs.a == 'hello'" },
          },
          blocks: [_.cloneDeep(variable1)],
        },
        {
          id: 'branch_f7fdf2437b4',
          type: 'case',
          definition: {
            info: { title: 'case_hello1' },
          },
          data: {
            condition: { type: 'expression', content: "variable.outputs.a == 'hello1'" },
          },
          blocks: [_.cloneDeep(variable1)],
        },
        {
          id: 'default_64e0c461509',
          type: 'default',
          definition: {
            info: {
              title: 'case_default',
              description: 'Executed when all conditions are not met',
              icon: '',
            },
          },
          blocks: [_.cloneDeep(variable2)],
        },
      ],
    }

    expect(translateNode(node)).toMatchSnapshot()
  })

  it('Condition - case(condition(condition))/default(condition)', () => {
    const node = _.cloneDeep(conditionMin) as FlowConditionNodeSchema

    node.blocks[0].blocks = [_.cloneDeep(conditionMin)]
    node.blocks[1].blocks = [_.cloneDeep(conditionMin)]
    node.blocks[2].blocks = [_.cloneDeep(conditionMin)]

    expect(translateNode(node)).toMatchSnapshot()
  })
})
