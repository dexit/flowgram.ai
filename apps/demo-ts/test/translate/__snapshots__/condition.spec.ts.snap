// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`translate Condition - case(condition(condition))/default(condition) 1`] = `
"  //=== condition@condition | {level:1,blockIndex:0}
  const condition = {}
  condition.outputs = {}
  //=== case_hello@case | {level:1,blockIndex:0,parent:'condition'}
  if (variable.outputs.a == 'hello') {
    //=== condition@condition | {level:2,blockIndex:0,parent:'case_hello'}
    const condition = {}
    condition.outputs = {}
    //=== case_hello@case | {level:2,blockIndex:0,parent:'condition'}
    if (variable.outputs.a == 'hello') {
    } // case_hello@case
    //=== case_hello1@case | {level:2,blockIndex:1,parent:'condition'}
    else if (variable.outputs.a == 'hello1') {
    } // case_hello1@case
    //=== case_default@default | {level:2,blockIndex:2,parent:'condition'}
    else {
    } // case_default@default
    } // case_hello@case
  //=== case_hello1@case | {level:1,blockIndex:1,parent:'condition'}
  else if (variable.outputs.a == 'hello1') {
    //=== condition@condition | {level:2,blockIndex:0,parent:'case_hello1'}
    const condition = {}
    condition.outputs = {}
    //=== case_hello@case | {level:2,blockIndex:0,parent:'condition'}
    if (variable.outputs.a == 'hello') {
    } // case_hello@case
    //=== case_hello1@case | {level:2,blockIndex:1,parent:'condition'}
    else if (variable.outputs.a == 'hello1') {
    } // case_hello1@case
    //=== case_default@default | {level:2,blockIndex:2,parent:'condition'}
    else {
    } // case_default@default
    } // case_hello1@case
  //=== case_default@default | {level:1,blockIndex:2,parent:'condition'}
  else {
    //=== condition@condition | {level:2,blockIndex:0,parent:'case_default'}
    const condition = {}
    condition.outputs = {}
    //=== case_hello@case | {level:2,blockIndex:0,parent:'condition'}
    if (variable.outputs.a == 'hello') {
    } // case_hello@case
    //=== case_hello1@case | {level:2,blockIndex:1,parent:'condition'}
    else if (variable.outputs.a == 'hello1') {
    } // case_hello1@case
    //=== case_default@default | {level:2,blockIndex:2,parent:'condition'}
    else {
    } // case_default@default
    } // case_default@default
  "
`;

exports[`translate Condition - case/case/default 1`] = `
"  //=== condition@condition | {level:1,blockIndex:0}
  const condition = {}
  condition.outputs = {}
  //=== case_hello@case | {level:1,blockIndex:0,parent:'condition'}
  if (variable.outputs.a == 'hello') {
    //=== variable_1@variable | {level:2,blockIndex:0,parent:'case_hello'}
    const variable_1 = {}
    variable_1.outputs = {}
    variable.outputs.a = 'hello world'
    } // case_hello@case
  //=== case_hello1@case | {level:1,blockIndex:1,parent:'condition'}
  else if (variable.outputs.a == 'hello1') {
    //=== variable_1@variable | {level:2,blockIndex:0,parent:'case_hello1'}
    const variable_1 = {}
    variable_1.outputs = {}
    variable.outputs.a = 'hello world'
    } // case_hello1@case
  //=== case_default@default | {level:1,blockIndex:2,parent:'condition'}
  else {
    //=== variable_2@variable | {level:2,blockIndex:0,parent:'case_default'}
    const variable_2 = {}
    variable_2.outputs = {
      a: '',
    }
    variable_2.outputs.a = start.outputs.query
    } // case_default@default
  "
`;

exports[`translate Condition - min 1`] = `
"  //=== condition@condition | {level:1,blockIndex:0}
  const condition = {}
  condition.outputs = {}
  //=== case_hello@case | {level:1,blockIndex:0,parent:'condition'}
  if (variable.outputs.a == 'hello') {
  } // case_hello@case
  //=== case_hello1@case | {level:1,blockIndex:1,parent:'condition'}
  else if (variable.outputs.a == 'hello1') {
  } // case_hello1@case
  //=== case_default@default | {level:1,blockIndex:2,parent:'condition'}
  else {
  } // case_default@default
  "
`;
